futures-core/src/task/atomic_waker.rs

```diff
@@ -87,6 +89,37 @@ impl AtomicWaker {
     /// idea. Concurrent calls to `register` will attempt to register different
     /// tasks to be notified. One of the callers will win and have its task set,
     /// but there is no guarantee as to which caller will succeed.
+    ///
+    /// # Examples
+    ///
+    /// Here is how `register` is used when implementing a flag.
+    ///
+    /// ```
+    /// # use futures_core::{Future, Poll, Never};
+    /// # use futures_core::Async::*;
+    /// # use futures_core::task::{self, AtomicWaker};
+    /// # use std::sync::atomic::AtomicBool;
+    /// # use std::sync::atomic::Ordering::SeqCst;
+    /// struct Flag {
+    ///     waker: AtomicWaker,
+    ///     set: AtomicBool,
+    /// }
+    ///
+    /// impl Future for Flag {
+    ///     type Item = ();
+    ///     type Error = Never;
+    ///
+    ///     fn poll(&mut self, cx: &mut task::Context) -> Poll<(), Never> {
+    ///         self.waker.register(cx.waker());
```

Comment:
I'm not sure I understand this example, but shouldn't the `register` only be called in the `Pending` case?

Reply:
It has to be called *before* checking the value. So, if you were to do that, you'd need to check `set`, call `register`, then check `set` again.

Reply:
I see-- because otherwise you might miss a wakeup. Can you leave a comment explaining that here?

Reply:
Done